nab/detectors/goodness_of_fit/goodness_of_fit_detector.py

```diff
@@ -0,0 +1,156 @@
+  Relative Entropy statistic with multiple hypothesis as described in
+  Figure 1 of Wang, Chengwei, et al. "Statistical Techniques for Online
+  Anomaly Detection in Data Centers", Proceedings of the 8th ACM
+  international conference on Autonomic computing. ACM, 2011.
+
+  The algorithm is based on a hypothesis testing approach that compares
+  observed data against multiple null hypotheses, representing frequencies of
+  quantized data over a window. If the observed data is unseen and does not
+  agree with any existing hypothesis, it is declared anomolous and a new
+  hypothesis is created. Otherwise, it is declared non-anomolous, provided that
+  the accepted hypothesis occurs frequently enough. Decision to accept/reject a
+  null hypothesis is based on relative entropy compared against a threshold
+  of acceptable false negative probability determined by the chi-squared
+  distribution. Step-by-step details are given in code comments and parameters
+  have been tuned for best performance of NAB.
+  """
+
+  def __init__(self, *args, **kwargs):
+    super(GoodnessOfFitDetector, self).__init__(*args, **kwargs)
+
+    # Timeseries of the metric on which anomaly needs to be detected
+    self.util = []
+
+    # Number of bins into which util is to be quantized
+    self.N_bins = 5.0
+
+    # Window size
+    self.W = 52
+
+    # Threshold against which the test statistic is compared. It is set to
+    # the point in the chi-squared cdf with N-bins -1 degrees of freedom that
+    #  corresponds to 0.99.
+    self.T = stats.chi2.isf(0.01, self.N_bins - 1)
+
+    # Threshold to determine if hypothesis has occured frequently enough
+    self.c_th = 1
+
+    # Tracks the current number of null hypothesis
+    self.m = 0
+
+    # Step size in timeseries quantization
+    self.stepSize = (self.inputMax - self.inputMin)/self.N_bins
+
+    # List of lists where P[i] indicates the empirical frequency of the ith
+    # hypothesis.
+    self.P = []
+
+    # List where c[i] tracks the number of windows that agree with P[i]
+    self.c = []
+
+
+  def handleRecord(self, inputData):
+    """ Returns a list of [anomalyScore] that takes a binary value of 0 or 1.
+    The anomalyScore is determined based on the agreement of the observed data
+    with existing hypotheses that occur frequently enough. Threshold to
+    accept/reject a null hypothesis and declare an anomaly is determined by
+    comparing the relative entropy of the observed data and all null
+    hypothesis against the point on chi-squared distribution that
+    corresponds to 0.99 (probability of incorrectly rejecting a
+    null-hypothesis).
+    """
+
+    anomalyScore = 0.0
+    inputValue = inputData["value"]
+    self.util.append(inputValue)
+
+    #  This check is for files where self.inputMin == self.input max i.e
+    #  all data points are identical and stepSize is 0 e.g
+    #  artificalNoAnomaly/art_flatline.csv file. Every point in such files
+    #  is declared non-anomolous.
```

Comment:
This is not in the paper and i have implemented it on my own. Without this, the denominator becomes 0 while running it on art_flatline.csv file because the entire file has identical metric values so min and max are equal.